Changelog 160 (#139)
intellij-rust/intellij-rust.github.io: _posts/2021-11-22-changelog-160.markdown

---
layout: post
title: "IntelliJ Rust Changelog #160"
date: 2021-11-22 13:00:00 +0300
---

## New Features

> * [#7996] Enable [crates local index](https://github.com/intellij-rust/intellij-rust/issues/6463) on stable to provide analysis for dependencies inside `Cargo.toml` manifest. Currently, it offers completions for `crates.io` dependencies' names and versions, inspections for non-existent packages in the index, and automatically updates with cargo crates index updates

Reviewer: Maybe drop mention of `inspections for non-existent packages` since it's disabled by default for now?

> * [#8031] Add "Merge `if`s" intention

Reviewer: ```suggestion
* [#8031] Add `` Merge `if`s `` intention
```

  {% include gif-img.html path="/assets/posts/changelog-160/merge_if" w="700px" %}

> * [#8030] Add "Simplify if with constant condition" inspection which works for both always `true` and always `false` if-conditions. The "Redundant else" inspection now checks only irrefutable patterns.

Reviewer: ```suggestion
* [#8030] Add `Condition of 'if' expression is constant` inspection which works for both always `true` and always `false` if-conditions. The `Redundant else` inspection now checks only irrefutable patterns.
```

* [#8102] [#8042] Enable [`Change signature` refactoring](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#change-sign) when a type of a trait method parameter changes (by [@Kobzol])

* [#7840] The intention for removing braces from a single expression now also works for blocks that contain a single statement (by [@Kobzol])

* [#6929] Detection of illegal lifetime names (by [@Stzx])

* [#7719] Support the [E0451](https://doc.rust-lang.org/error-index.html#E0451) compiler error ("a struct constructor with private fields was invoked") for struct field literals using [shorthand init](https://doc.rust-lang.org/reference/expressions/struct-expr.html#struct-field-init-shorthand) (by [@Kobzol])

## Performance Improvements

* [#8089] Avoid using potentially long-running computation inside the UI thread when invoking the [`Introduce constant`](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-refactorings.html#extractconst-refactoring) refactoring (by [@Kobzol])

* [#8043] Do not use potentially long-running operations on the UI thread in the `Change signature` refactoring (by [@Kobzol])

## Fixes

> * Generated sources fixes:

Reviewer: ```suggestion
* [Build script](https://doc.rust-lang.org/cargo/reference/build-scripts.html) evaluation fixes:
```

>   * [#8012] Improve work with generated sources provided via `org.rust.cargo.evaluate.build.scripts` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features). Now the plugin doesn't pollute `Project Structure` view with entries for generated `out` directories. Also, the custom paths of target directories are also supported

Reviewer: ```suggestion
  * [#8012] Improve work with sources generated by build scripts. Now the plugin doesn't pollute `Project Structure` view with entries for generated `out` directories. Also, the custom paths of target directories are also supported
```

>   * [#8056] Fix build script evaluation in [workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html). Note, build script evaluation is disabled by default. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features)

Reviewer: We still write about `org.rust.cargo.evaluate.build.scripts` experimental feature twice. Can we lift this note into `Generated sources fixes` item and write it only once?

>   * [#8056] Fix build script evaluation in [workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html). Note, build script evaluation is disabled by default. To turn it on, enable `org.rust.cargo.evaluate.build.scripts` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features)

Reviewer: ```suggestion
  * [#8056] Fix build script evaluation in [workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html). Now, the plugin properly evaluates scripts in all packages of the workspace
```

Sorry, I had to write this addition in the PR at the beginning
